mingpt/mingptmlx.py

```diff
@@ -0,0 +1,309 @@
+    data = train_data if split == 'train' else val_data
+    ix = mx.random.randint(0, len(data) - block_size, (batch_size,))
+    ix = [i.item() for i in ix]
+    x = mx.stack([data[i:i+block_size] for i in ix])
+    y = mx.stack([data[i+1:i+block_size+1] for i in ix])
+    # x, y = x.to(device), y.to(device)
+    return x, y
+
+
+# %%
+
+class Head(nn.Module):
+    """ one head of self-attention """
+
+    def __init__(self, head_size):
+        super().__init__()
+        self.key = nn.Linear(n_embd, head_size, bias=False)
+        self.query = nn.Linear(n_embd, head_size, bias=False)
+        self.value = nn.Linear(n_embd, head_size, bias=False)
+        self.tril = mx.tril(mx.ones([block_size, block_size]))
+        self.dropout = nn.Dropout(dropout)
+
+    def __call__(self, x):
+        B,T,C = x.shape
+        k = self.key(x)   # (B,T,C)
+        q = self.query(x) # (B,T,C)
+        # compute attention scores ("affinities")
+        wei = q @ k.transpose((0,2,1)) * C**-0.5 # (B, T, C) @ (B, C, T) -> (B, T, T)
+        mask = self.tril[:T, :T] == 0
+        wei = mx.where(mask, float('-inf'), wei) # (B, T, T)
+        wei = nn.softmax(wei, axis=-1) # (B, T, T)
+        wei = self.dropout(wei)
+        # perform the weighted aggregation of the values
+        v = self.value(x) # (B,T,C)
+        out = wei @ v # (B, T, T) @ (B, T, C) -> (B, T, C)
+        return out
+
+
+# %%
+
+class MultiHeadAttention(nn.Module):
+    """ multiple heads of self-attention in parallel """
+
+    def __init__(self, num_heads, head_size):
+        super().__init__()
+        self.heads = [Head(head_size) for _ in range(num_heads)]
+        self.proj = nn.Linear(n_embd, n_embd)
+        self.dropout = nn.Dropout(dropout)
+
+    def __call__(self, x):
+        out = mx.concatenate([h(x) for h in self.heads], axis=-1)
+        out = self.dropout(self.proj(out))
+        return out
+
+class FeedFoward(nn.Module):
+    """ a simple linear layer followed by a non-linearity """
+
+    def __init__(self, n_embd):
+        super().__init__()
+        self.net = nn.Sequential(
+            nn.Linear(n_embd, 4 * n_embd),
+            nn.ReLU(),
+            nn.Linear(4 * n_embd, n_embd),
+            nn.Dropout(dropout),
+        )
+
+    def __call__(self, x):
+        return self.net(x)
+
+
+# %%
+
+class Block(nn.Module):
+    """ Transformer block: communication followed by computation """
+
+    def __init__(self, n_embd, n_head):
+        # n_embd: embedding dimension, n_head: the number of heads we'd like
+        super().__init__()
+        head_size = n_embd // n_head
+        self.sa = MultiHeadAttention(n_head, head_size)
+        self.ffwd = FeedFoward(n_embd)
+        self.ln1 = nn.LayerNorm(n_embd)
+        self.ln2 = nn.LayerNorm(n_embd)
+
+    def __call__(self, x):
+        x = x + self.sa(self.ln1(x))
+        x = x + self.ffwd(self.ln2(x))
+        return x
+
+
+# %%
+# super simple bigram model
+class BigramLanguageModel(nn.Module):
```

Comment:
Why is it called a `BigramLanguageModel`?

Reply:
Didn’t think of it until now. Good question. It looks like a name that came from a previous video Karpathy made where the model was specifically looking at two chars to make a predication on next char. https://github.com/karpathy/makemore/blob/master/makemore.py#L411

In the context of this minGPT though I don’t see how we’re looking at two chars (or two-grams) anywhere.  Perhaps the name doesn’t make sense. Perhaps SimpleLanguageModel is more apt.